### PR TITLE
1695 add empty state component

### LINKED ISSUE #1695: [Empty state] Empty states as an independent component
### What package is this for?

- [x] React
- [ ] Angular

Build empty states as it's own component so that it can be used in different contexts vs being created for each use case separately.

This is based on the [Carbon empty state pattern](https://www.carbondesignsystem.com/patterns/empty-states-pattern), but we have created AI Apps specific [empty state guidance](https://pages.github.ibm.com/ai-applications/design/components/empty-states/usage) that includes content suggestions and six illustrations that can universally be used.

**Note:** These empty states work for full pages and larger spaces such as lists and tables. The empty states for smaller spaces such as dashboard cards might be unique (with no image).

image

Illustration SVGs can be downloaded [here](https://ibm.box.com/s/x224l8t2dlqg0p0w3sxsu1piqqalpm5j)

Design specs:
Empty state specs

### PR DESCRIPTION
**Pull request by jonguenther**
Closes #1695 

**Summary**

- Adds `EmptyState` component according to the carbon / AI-PAL guidelines

**Change List (commits, features, bugs, etc)**

- Add `EmpyState` component
- Add `EmptyState` icons for the different types of empty states
- Add story for component
- Add tests for component

**Acceptance Test (how to verify the PR)**

- View [the stories](https://deploy-preview-1793--carbon-addons-iot-react.netlify.app/?path=/story/watson-iot-emptystate--first-time-use), they should match [the design site guidance](https://pages.github.ibm.com/ai-applications/design/components/empty-states/usage/)
- checks should pass

### CODE REVIEW

**Review by tay1orjones — CHANGES_REQUESTED** [review]
Overall looks good! A few comments below on the code. I'm going to push a couple small changes to your stories if that's alright with you.

**Comment by tay1orjones** [comment]
In addition, based on guidance from @dianatran18, the actions shouldn't be fully customizable. We should only allow the buttons, not an open ended node/element.

File: src/components/EmptyState/EmptyState.test.jsx [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+  //   expect(screen.getByTestId('customcomponent')).toBeTruthy();
+  // });
+
+  // it('shows custom secondaryAction component', () => {
+  //   const customAction = <div data-testid="customcomponent">Hello</div>;
+  //   render(<EmptyState {...commonProps} secondaryAction={customAction} />);
+
+  //   expect(screen.getByTestId('customcomponent')).toBeTruthy();
+  // });
```
Comment (tay1orjones) [inline src/components/EmptyState/EmptyState.test.jsx:106]:
```suggestion
```

File: src/components/EmptyState/EmptyState.jsx [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            </Link>
+          )}
+          {/* {secondaryAction.label ? (
+            // eslint-disable-next-line jsx-a11y/anchor-is-valid
+            <Link onClick={secondaryAction.onClick && secondaryAction.onClick}>
+              {secondaryAction.label}
+            </Link>
+          ) : (
+            secondaryAction
+          )} */}
```
Comment (tay1orjones) [inline src/components/EmptyState/EmptyState.jsx:150]:
```suggestion
```

File: src/components/EmptyState/EmptyState.jsx [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+          <Button onClick={action.onClick && action.onClick}>
+            {action.label}
+          </Button>
+          {/* {action.label ? (
+            <Button onClick={action.onClick && action.onClick}>
+              {action.label}
+            </Button>
+          ) : (
+            action
+          )} */}
```
Comment (tay1orjones) [inline src/components/EmptyState/EmptyState.jsx:130]:
```suggestion
```

File: src/components/EmptyState/EmptyState.jsx [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    label: PropTypes.string.isRequired,
+    onClick: PropTypes.func.isRequired,
+  }),
+  // secondaryAction: PropTypes.oneOfType([
+  //   PropTypes.func,
+  //   PropTypes.shape({
+  //     label: PropTypes.string.isRequired,
+  //     onClick: PropTypes.func.isRequired,
+  //   }),
+  // ]),
```
Comment (tay1orjones) [inline src/components/EmptyState/EmptyState.jsx:69]:
```suggestion
```

File: src/components/EmptyState/EmptyState.jsx [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    label: PropTypes.string.isRequired,
+    onClick: PropTypes.func.isRequired,
+  }),
+  // action: PropTypes.oneOfType([
+  //   PropTypes.func,
+  //   PropTypes.shape({
+  //     label: PropTypes.string.isRequired,
+  //     onClick: PropTypes.func.isRequired,
+  //   }),
+  // ]),
```
Comment (tay1orjones) [inline src/components/EmptyState/EmptyState.jsx:57]:
```suggestion
```

File: src/components/EmptyState/EmptyState.jsx [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+const icons = {
+  error: ErrorImage,
+  error404: Error404Image,
+  empty: EmptyImage,
+  'not-authorized': NotAuthImage,
+  'no-result': NoResultImage,
+  success: SuccessImage,
+};
+
+// TODO: Discuss whether actions can be custom components, e.g. for showing details in error messages.
```
Comment (tay1orjones) [inline src/components/EmptyState/EmptyState.jsx:27]:
```suggestion
```

**Review by tay1orjones — COMMENTED** [review]
Some comments to remove, but otherwise looks great - thanks!

**Review by tay1orjones — APPROVED** [review]
Thanks again for this contribution! 🎉

File: src/components/EmptyState/EmptyState.jsx [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+      <p
+        className={`${iotPrefix}--empty-state--text`}
+        data-testid={`${testID}-body`}>
+        {body}
+      </p>
+      {action && (
+        <div
+          className={`${iotPrefix}--empty-state--action`}
+          data-testid={`${testID}-action`}>
+          <Button onClick={action.onClick && action.onClick}>
```
Comment (bjornalm) [inline src/components/EmptyState/EmptyState.jsx:106]: Why `action.onClick && action.onClick`? 
(We know there is an action and if there is an action then the `onClick` prop is required)

File: src/components/EmptyState/EmptyState.jsx [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+          </Button>
+        </div>
+      )}
+      {secondaryAction && (
+        <div
+          className={`${iotPrefix}--empty-state--link`}
+          data-testid={`${testID}-secondaryAction`}>
+          {secondaryAction.label && (
+            // eslint-disable-next-line jsx-a11y/anchor-is-valid
+            <Link onClick={secondaryAction.onClick && secondaryAction.onClick}>
```
Comment (bjornalm) [inline src/components/EmptyState/EmptyState.jsx:117]: See action.onClick comment

**Review by bjornalm — APPROVED** [review]
Nice work, just a couple of comments.

**Pull request merged**